### PR TITLE
Use `Span`s to identify unreachable subpatterns in or-patterns

### LINKED ISSUE #71977: or_patterns: incorrect unreachable pattern warning
```rust
#![feature(or_patterns)]

enum F {
    A,
    B,
}

fn main() {
    let a = F::A;
    let b = F::B;
    match (a, b) {
        (F::A, F::A) => (),
        (F::B | F::A, F::B | F::A) => (),
    }
}
```
Results in the following incorrect warning:
```
warning: unreachable pattern
  --> src/main.rs:13:30
   |
13 |         (F::B | F::A, F::B | F::A) => (),
   |                              ^^^^
   |
   = note: `#[warn(unreachable_patterns)]` on by default

warning: 1 warning emitted
```
Removing `F::A` causes a non-exhaustive patterns error as expected.

[playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=c4e3fbb48fc2d34cd43bc0896ce01940)

### PR DESCRIPTION
**Pull request by Nadrieril**
Fixes #71977

### CODE REVIEW

File: src/librustc_mir_build/hair/pattern/_match.rs [hunk 1]
```diff
@@ -1852,16 +1853,35 @@ crate fn is_useful<'p, 'tcx>(
         // We need to push the already-seen patterns into the matrix in order to detect redundant
         // branches like `Some(_) | Some(0)`. We also keep track of the unreachable subpatterns.
         let mut matrix = matrix.clone();
-        let mut unreachable_pats = Vec::new();
+        // Of all the branches of the current or-pattern, which are unreachable.
```
Comment (teymour-aldridge) [inline src/librustc_mir_build/hair/pattern/_match.rs:1856]:
```suggestion
        // `Vec` of all the unreachable branches of the current or-pattern.
```

File: src/test/ui/or-patterns/exhaustiveness-unreachable-pattern.rs [hunk 2]
```diff
@@ -67,4 +70,29 @@ fn main() {
              | 1) => {}
         _ => {}
     }
+
+    // A subpattern that is unreachable in one branch only is overall reachable.
```
Comment (teymour-aldridge) [inline src/test/ui/or-patterns/exhaustiveness-unreachable-pattern.rs:74]:
```suggestion
    // A subpattern that is only unreachable in one branch is overall reachable.
```

**Comment by matthewjasper** [comment]
Can you remove the `allow(unreachable_patterns)` from src/test/ui/or-patterns/search-via-bindings.rs?

**Comment by Nadrieril** [comment]
@matthewjasper I removed it and indeed it's not needed anymore

**Comment by Manishearth** [comment]
Please use descriptive PR titles

**Pull request merged**